apps/src/javalab/JavalabCaptchaDialog.jsx

```diff
@@ -0,0 +1,86 @@
+import React from 'react';
+import {connect} from 'react-redux';
+import PropTypes from 'prop-types';
+import javalabMsg from '@cdo/javalab/locale';
+import {
+  appendNewlineToConsoleLog,
+  appendOutputLog,
+} from '@cdo/apps/javalab/redux/consoleRedux.ts';
+import {setIsCaptchaDialogOpen} from '@cdo/apps/javalab/redux/javalabRedux.ts';
+import SafeMarkdown from '@cdo/apps/templates/SafeMarkdown';
+import ReCaptchaDialog from '@cdo/apps/templates/ReCaptchaDialog';
+
+const BLOG_URL = 'https://support.code.org/hc/en-us/articles/6104078305549';
+
+// exported for tests
+export function UnconnectedJavalabCaptchaDialog({
+  onVerify,
+  onCancel,
+  isCaptchaDialogOpen,
+  setIsCaptchaDialogOpen,
+  appendNewlineToConsoleLog,
+  appendOutputLog,
+  recaptchaSiteKey,
+}) {
+  // return promise for tests
+  const onCaptchaSubmit = token => {
+    return fetch('/dashboardapi/v1/users/me/verify_captcha', {
```

Comment:
Just curious - it's interesting that the `RecaptchaDialog` itself doesn't already just make this call? Do you know if we have other/multiple recaptcha endpoints?

Reply:
Yeah the `RecaptchaDialog` is not in use at the moment (and explicitly tells you to make your own server call). I think the more standard way to use captchas is in a form, which is what we do at studio.code.org/report_abuse:

https://github.com/code-dot-org/code-dot-org/blob/5a7df81de4108acf7158ce2339516dfc43e95422/apps/src/code-studio/components/ReportAbuseForm.jsx#L169-L175

Google loads the captcha into the form (wherever the special `g-recaptcha` class name is) -- it submits a google generated token as part of the form submission, which can be validated in a controller using this magic `verify_captcha` method:

https://github.com/code-dot-org/code-dot-org/blob/5a7df81de4108acf7158ce2339516dfc43e95422/dashboard/app/controllers/report_abuse_controller.rb#L37-L41

So, we definitely use captcha in a one or a couple other places, but they're typically less about an explicit captcha verification (and storage of a timestamp when verification happened) and more used to guard access to some endpoint.